src/script/descriptor.cpp

```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
     BIP32PubkeyProvider(const CExtPubKey& extkey, KeyPath path, DeriveType derive) : m_extkey(extkey), m_path(std::move(path)), m_derive(derive) {}
     bool IsRange() const override { return m_derive != DeriveType::NO; }
     size_t GetSize() const override { return 33; }
-    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info) const override
+    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) const override
     {
+        KeyOriginInfo info_out, parent_info;
+        CKeyID keyid = m_extkey.pubkey.GetID();
+        std::copy(keyid.begin(), keyid.begin() + sizeof(info_out.fingerprint), info_out.fingerprint);
+        info_out.path = m_path;
+        parent_info = info_out;
+        if (m_derive == DeriveType::UNHARDENED) info_out.path.push_back((uint32_t)pos);
+        if (m_derive == DeriveType::HARDENED) info_out.path.push_back(((uint32_t)pos) | 0x80000000L);
         if (key) {
-            if (IsHardened()) {
-                CKey priv_key;
-                if (!GetPrivKey(pos, arg, priv_key)) return false;
-                *key = priv_key.GetPubKey();
+            CExtPubKey extkey = m_extkey;
+            CExtPubKey parent = extkey;
```

Comment:
This `CExtPubKey` juggling still confuses me :-)

I added some comments for my own sanity. Will probably need to read this again later.

```
CExtPubKey extkey_out = m_extkey;    // to hold extended key for *key (m/.../k)
CExtPubKey extkey_parent = m_extkey; // to hold extended key for the parent of *key (m/.../k/pos)
```